### PR TITLE
Add generateConfig action, updates to check-update-security and generate-release-vars.sh

### PR DESCRIPTION
**Pull request by hozkaya2000**
### Summary
Add generate config github action. The action checks runs ./generate-release-vars.sh, sees if there are any dependency updates, and pushes to the repository in the case of a release needing to be kicked off. If no dependency updates exist, it will fall back to checking security updates by launching an ec2 instance with the AMIs and checking for critical updates.

Also includes a couple modifications to the check update security script to make it more resilient to API errors, that occurred in a couple rare test scenarios. 

### Implementation details
The action was split into steps of a github action, and runs on workflow dispatch as of now. 

### Testing
A testing document was made to check all important test scenarios. 

New tests cover the changes: yes

### Description for the changelog
Add GenerateConfig github action

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/generateconfig.yml [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    - name: Configure prereqs
+      run: |
+        git config --global user.name "GenerateConfig Action"
+        git config --global user.email "gcaction@github.com"
+        UpdateAl1="false"
+        UpdateAl2="false"
+        UpdateAl2023="false"
+    - name: Check AL1 Updates
+      run: |
+        source_ami_line_nums=(2 3 4 5)
```
Comment (sparrc) [inline .github/workflows/generateconfig.yml:32]: could we put all this bash in the repo `scripts/` folder rather than having it defined in in-line yaml?

Comment (sparrc) [reply .github/workflows/generateconfig.yml:32]: so that in yaml this just becomes something like `run: scripts/generate-al1-config` or `run: scripts/generate-config al1`

Comment (hozkaya2000) [reply .github/workflows/generateconfig.yml:32]: Now that you mention it, I think this is a good idea. I will do so.

File: .github/workflows/generateconfig.yml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        source_ami_line_contents=()
+        al1_file_path="./release-al1.auto.pkrvars.hcl"
+        for line_number in "${source_ami_line_nums[@]}"; do
```
Comment (sparrc) [inline .github/workflows/generateconfig.yml:35]: this seems a bit complicated parsing logic, would it be easier to have our "canonical" variables defined in json and then generate the .hcl file from that?

File: .github/workflows/generateconfig.yml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        role-to-assume: ${{secrets.AMI_GENERATE_CONFIG_ROLE}}
+        aws-region: us-west-2
+    - name: Configure prereqs
+      run: |
+        git config --global user.name "GenerateConfig Action"
+        git config --global user.email "gcaction@github.com"
+        UpdateAl1="false"
+        UpdateAl2="false"
+        UpdateAl2023="false"
+    - name: Check AL1 Updates
```
Comment (amogh09) [inline .github/workflows/generateconfig.yml:30]: I don't quite understand how this job works. Can you please explain a little?

File: check-update-security.sh [hunk 4]
```diff
@@ -89,7 +89,7 @@ esac
 # Query ssm to get latest ecs optimized ami
 ami_id=$(aws ssm get-parameters --names $ami_path --region us-west-2 | jq -r '.Parameters[0].Value' | jq -r '.image_id')
 
-user_data=$(mktemp user_data.txt)
+user_data=$(touch user_data.txt)
```
[the remark this answers is not in the document]
Comment (hozkaya2000) [reply check-update-security.sh:92]: mktemp was giving some errors regarding requiring a certain number of 'X's in the file name when testing, so I switched it to making the file normally, since we delete it afterwards anyway. Also, since the environment where github actions is being run is cleaned and shut off after the action is finished, the file need not be deleted later.

File: check-update-security.sh [hunk 5]
```diff
@@ -153,7 +153,7 @@ check_wait_response $(echo $?)
 # Instance has been launched, terminate in case of an error
 trap 'failure_cleanup' ERR
 
-rm "$user_data"
+rm user_data.txt
```
[the remark this answers is not in the document]
Comment (hozkaya2000) [reply check-update-security.sh:156]: because we went from mktemp to touch as the action, touch doesn't return the name of the file, just a success code. Therefore, we can't try to remove "0" or anything like that. We have to actually remove the file.

File: check-update-security.sh [hunk 6]
```diff
@@ -195,12 +195,12 @@ command_status() {
 max_retries=20
 success=0
 for ((r = 0; r < max_retries; r++)); do
+    sleep 5
```
[the remark this answers is not in the document]
Comment (hozkaya2000) [reply check-update-security.sh:198]: This was done because there was a very small corner case where the command invocation is seems to not exist at the first check. Instead of risking that chance, we wait 5 seconds. None of the tests after this change procured such the same issue.

[the remark this answers is not in the document]
Comment (amogh09) [reply check-update-security.sh:198]: You mean `aws` command does not exist on the host but then appears later?

File: scripts/check-update.sh [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    for ((i = 0; i < "${#source_ami_line_nums[@]}"; i++)); do
+        line_number=${source_ami_line_nums[$i]}
+        previous_line_content=${source_ami_line_contents[$i]}
+        updated_line_content=$(sed -n "${line_number}p" "$al1_file_path")
+        if [ "$updated_line_content" != "$previous_line_content" ]; then
+            UpdateAl1="true"
+            break
+        fi
+    done
```
[the remark this answers is not in the document]
Comment (hozkaya2000) [reply scripts/check-update.sh:44]: So line 1 is guaranteed to be changed every time. Sadly, that's why we have to check line by line. We also don't want to keep that line intact in the generate-release-vars because that will make it so multiple different scripts are changing the same file, which can be harder to control when testing.

File: .github/workflows/generateconfig.yml (hunk 1 above)
Comment (hozkaya2000) [reply .github/workflows/generateconfig.yml]: done.

File: scripts/check-update.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (amogh09) [reply scripts/check-update.sh:44]: We can also save the old contents to a temporary file and do something like 
```
diff <(grep -v ami_version release.old) <(grep -v ami_version release.new)
```

File: check-update-security.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (hozkaya2000) [reply check-update-security.sh:198]: nope, I mean when we invoke the _command on the instance_ by ssm, trying to check the _command invocation_ using comand_status() right away gives an error of not being able to find such an invocation on the instance. Perhaps it is not completely processed when the first aws call returns and not available until a few seconds pass.

File: scripts/check-update.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (hozkaya2000) [reply scripts/check-update.sh:44]: I see. Let me try this out real quick and try to add it in.

File: check-update-security.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (amogh09) [reply check-update-security.sh:198]: Okay thanks for explaining.

File: scripts/check-update.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (hozkaya2000) [reply scripts/check-update.sh]: Thanks for the elegant solution amogh, its now implemented instead of the line by line solution if you'd like to check it out.

File: scripts/check-update.sh [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        UpdateAl1="true"
+    fi
+
+    if [ "$UpdateAl1" != "true" ]; then
+        UpdateAl1=$(./check-update-security.sh al1)
+    fi
+    if [ "$UpdateAl1" = "true" ]; then
+        git add release-al1.auto.pkrvars.hcl
+        echo "Al1 Has Updates"
+    fi
```
Comment (amogh09) [inline scripts/check-update.sh:41]: Can the two if/else blocks be merged?

File: scripts/check-update.sh [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    if [ "$UpdateAl2" != "true" ]; then
+        UpdateAl2=$(./check-update-security.sh al2)
+        if [ "$UpdateAl2" != "true" ]; then
+            UpdateAl2=$(./check-update-security.sh al2_arm)
+        fi
+    fi
+    if [ "$UpdateAl2" = "true" ]; then
+        git add release-al2.auto.pkrvars.hcl
+        echo "Al2 Has Updates"
+    fi
```
Comment (amogh09) [inline scripts/check-update.sh:62]: Do you think we can capture these actions in a function and call it three times for AL1, AL2, and AL2023?

File: scripts/check-update.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (amogh09) [reply scripts/check-update.sh]: 💯

File: scripts/check-update.sh (hunk 9 above)
Comment (hozkaya2000) [reply scripts/check-update.sh]: I modified it so its much cleaner now, thanks for pointing this out. Instead of a case by case basis, the file now runs the whole thing for the specific AMI with the same code.

File: .github/workflows/generateconfig.yml (hunk 2 above)
Comment (hozkaya2000) [reply .github/workflows/generateconfig.yml]: I see. However, since I changed it such that we're looking at the diff except the line with ami_version with Amogh's suggestion, I think it is a bit cleaner now and doesn't include a line by line check.

**Review by amogh09 — APPROVED** [review]

**Review by sparrc — APPROVED** [review]
